distribution/client/src/lib/snapshotter.js

```diff
@@ -0,0 +1,127 @@
+/*
+* Used for snapshotting and rollbacks during upgrades.
+* See:
+*  https://github.com/jenkinsci/jep/tree/master/jep/302
+*/
+
+const { spawnSync } = require('child_process');
+const logger        = require('winston');
+const fs            = require('fs');
+
+const LOG_PREFIX = '[snapshotting]';
+// https://github.com/jenkinsci/jep/tree/master/jep/302#user-content-files-to-store
+const GITIGNORE_CONTENT = `
+/plugins/
+/updates/
+/secrets/master.key
+`;
+
+class Snapshotter {
+
+  constructor(app, options) {
+    this.options = options || {};
+  }
+
+  init(workingDirectory) {
+    if (!workingDirectory) {
+      throw new Error('workingDirectory parameter is required');
+    }
+    this.workingDirectory = workingDirectory;
+
+    if (!fs.existsSync(this.getDotGitDirectory())) {
+      this.git('init');
+      this.git('config', 'user.email', 'evergreen-client@jenkins.io');
+      this.git('config', 'user.name', '"Evergreen Client Snapshotting System (JEP 302)"');
+      this.git('commit','--allow-empty', '--message', 'Initial evergreen commit');
+
+      fs.writeFile(`${this.getDotGitDirectory()}/description`,
+        `This repository is used and specifically designed for Evergreen snapshotting system
+    (https://github.com/jenkinsci/jep/tree/master/jep/302).
+    It is *not* intended to act as a backup system in any way.
+    Never push this repository outside of the Evergreen container to avoid any risk of
+    leaking sensitive data.`,
+        (err) => {
+          if (err) {
+            logger.error('Impossible to write the repository description. Abnormal but not blocking.', err);
+          }
+        });
+    } else {
+      logger.info(`${LOG_PREFIX} ${this.getDotGitDirectory()} already initialized, moving on`);
+    }
+    this.updateGitIgnore();
+  }
+
+  /**
+   * https://github.com/jenkinsci/jep/tree/master/jep/302#take-a-snapshot
+   */
+  // TODO: design what needs to be passed (UL from/to, etc.)
+  snapshot(message) {
+    if (!this.workingDirectory) {
+      throw new Error('init() must be called first');
+    }
+
+    this.updateGitIgnore();
+
+    this.git('add', '--all');
+    // TODO: allow-empty but should trigger a warning if nothing was changed
+    this.git('commit', '--allow-empty', '--message', `${message}`);
+  }
+
+  /**
+   * Will update the .gitignore file and commit if needed.
+   * No-Op if already containing the expected content.
+   */
+  updateGitIgnore() {
```

Comment:
I'm wondering if the `.gitignore` file should simply be part of the distribution that we ship? Do you think it will change over time?

Reply:
Well, I did give this some thought, and this looked easier for the same outcome to put this in the code directly. As we're able to update the js client code anyway, this looked the same to me as referring to an updatable `.gitignore` file from the disk.